### PR TITLE
feat: configure BGE embedding profile

### PR DESCRIPTION
**Pull request by hellkonig**
## Summary
- add configurable pooling, normalization, and max-length vectorizer settings
- resolve BGE-small auto settings to CLS pooling with L2 normalization and reject incorrect explicit BGE configs
- version embedding cache keys by resolved embedding configuration
- update example config and README to use BGE-small

## Verification
- uv run ruff check .
- uv run ruff format --check .
- uv run python -m mypy src
- uv run python -m pytest

### CODE REVIEW

File: src/arxiv_recommender/schemas/config.py [hunk 1]
```diff
@@ -1,7 +1,12 @@
+from typing import Literal
+
 from pydantic import BaseModel, Field, field_validator
 
 from arxiv_recommender.utils.logging import SUPPORTED_LOG_LEVELS
 
+EmbeddingPoolingStrategy = Literal["auto", "mean", "cls"]
+NormalizeEmbeddings = bool | Literal["auto"]
+
```
Comment (hellkonig) [inline src/arxiv_recommender/schemas/config.py]: Can you revisit this to see if it follows the best practise of the high coding standard

File: src/arxiv_recommender/text_vectorization/cache.py [hunk 2]
```diff
@@ -36,7 +38,8 @@ def _compute_key(self, text: str) -> str:
         Returns:
             str: SHA256 hash of the text.
         """
-        return hashlib.sha256(text.encode()).hexdigest()
+        cache_input = f"{self.namespace}\n{text}"
+        return hashlib.sha256(cache_input.encode()).hexdigest()
```
Comment (hellkonig) [inline src/arxiv_recommender/text_vectorization/cache.py:42]: why do you need sha? the string should be able to be the key in a dictionary

File: src/arxiv_recommender/text_vectorization/huggingface_embed.py [hunk 3]
```diff
@@ -7,27 +7,53 @@
...
+ResolvedPoolingStrategy = Literal["mean", "cls"]
+NormalizeEmbeddings = bool | Literal["auto"]
+
+MODEL_EMBEDDING_PROFILES: dict[str, dict[str, ResolvedPoolingStrategy | bool]] = {
+    "BAAI/bge-small-en-v1.5": {
+        "pooling_strategy": "cls",
+        "normalize_embeddings": True,
+    },
+}
+
```
Comment (hellkonig) [inline src/arxiv_recommender/text_vectorization/huggingface_embed.py]: Do you think writing them here follows the best practise of the code?

File: src/arxiv_recommender/text_vectorization/huggingface_embed.py [hunk 4]
```diff
@@ -68,28 +94,63 @@ def tokenize(self, text: str) -> dict[str, torch.Tensor]:
...
 
-        embeddings = outputs.last_hidden_state
+        embeddings = cast(torch.Tensor, outputs.last_hidden_state)
         attention_mask = tokenized_text["attention_mask"]
-        return self._mean_pool(embeddings, attention_mask)
+        if self.pooling_strategy == "cls":
+            pooled_embeddings = embeddings[:, 0]
+        else:
+            pooled_embeddings = self._mean_pool(embeddings, attention_mask)
+
```
Comment (hellkonig) [inline src/arxiv_recommender/text_vectorization/huggingface_embed.py:107]: Should we use a different way (e.g. factory pattern) to load the pooling strategy?

File: src/arxiv_recommender/text_vectorization/huggingface_embed.py [hunk 5]
```diff
@@ -68,28 +94,63 @@ def tokenize(self, text: str) -> dict[str, torch.Tensor]:
...
     def _mean_pool(self, embeddings: torch.Tensor, attention_mask: torch.Tensor) -> torch.Tensor:
         mask = attention_mask.unsqueeze(-1).expand(embeddings.size()).float()
         masked_embeddings = embeddings * mask
         summed_embeddings = torch.sum(masked_embeddings, dim=1)
         token_counts = torch.clamp(mask.sum(dim=1), min=1e-9)
-        sentence_embeddings = summed_embeddings / token_counts
-        return torch.mean(sentence_embeddings, dim=0, keepdim=False)
+        return summed_embeddings / token_counts
+
+    def _resolve_embedding_config(
```
Comment (hellkonig) [inline src/arxiv_recommender/text_vectorization/huggingface_embed.py]: is there an elegant way to resolve it? The current code is working, but it doesn't look good.

File: tests/text_vectorization/test_huggingface_embed.py [hunk 6]
```diff
@@ -62,6 +62,101 @@ def test_process_uses_cache(
...
+                "BAAI/bge-small-en-v1.5",
+                pooling_strategy="mean",
+                normalize_embeddings=True,
+            )
+
+    def test_bge_small_rejects_incorrect_explicit_normalization(self) -> None:
+        with self.assertRaisesRegex(
+            ValueError,
+            "BAAI/bge-small-en-v1.5 requires pooling_strategy='cls'",
+        ):
```
Comment (hellkonig) [inline tests/text_vectorization/test_huggingface_embed.py:108]: Why does the error say the model requires pooling_strategy = 'cls'? Here we test the normalisation issue.

File: src/arxiv_recommender/schemas/config.py (hunk 1 above)
Comment (hellkonig) [reply src/arxiv_recommender/schemas/config.py]: Addressed in 90624fe. I moved the shared embedding configuration types and model-profile resolver into a dedicated top-level embedding_config module, so the schema no longer owns ad hoc Literal aliases and the resolver remains reusable by vectorization code.

File: src/arxiv_recommender/text_vectorization/huggingface_embed.py (hunk 3 above)
Comment (hellkonig) [reply src/arxiv_recommender/text_vectorization/huggingface_embed.py]: Addressed in 90624fe. The BGE profile registry is no longer in huggingface_embed.py; it now lives in the dedicated embedding_config module with the resolver and profile dataclasses.

File: src/arxiv_recommender/text_vectorization/cache.py (hunk 2 above)
Comment (hellkonig) [reply src/arxiv_recommender/text_vectorization/cache.py:42]: I kept the SHA key intentionally. The cache already used SHA before this PR; this change only added the embedding namespace to the hashed input. A raw string or tuple key would work, but hashing avoids storing full paper text in the OrderedDict keys and keeps key size fixed. Collision risk is negligible for this in-memory cache.

File: src/arxiv_recommender/text_vectorization/huggingface_embed.py (hunk 4 above)
Comment (hellkonig) [reply src/arxiv_recommender/text_vectorization/huggingface_embed.py:107]: Addressed in 90624fe with a small pooler factory instead of a larger class hierarchy. huggingface_embed.py now calls create_pooler(...) and does not branch on pooling strategy inside vectorize().

File: src/arxiv_recommender/text_vectorization/huggingface_embed.py (hunk 5 above)
Comment (hellkonig) [reply src/arxiv_recommender/text_vectorization/huggingface_embed.py]: Addressed in 90624fe. The resolver is now a standalone resolve_embedding_config(...) function returning a frozen ResolvedEmbeddingConfig. HuggingFaceEmbedding just consumes the resolved config and builds the cache namespace from it.

File: tests/text_vectorization/test_huggingface_embed.py (hunk 6 above)
Comment (hellkonig) [reply tests/text_vectorization/test_huggingface_embed.py:108]: Fixed in 90624fe. The normalization test now expects a normalization-specific error message: normalize_embeddings must be True, got False.

File: src/arxiv_recommender/embedding_config.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    if profile:
+        _validate_known_model_config(model_name, profile, resolved_config)
+
+    return resolved_config
+
+
+def _coerce_normalize_embeddings(value: NormalizeEmbeddings | str) -> NormalizeEmbeddings:
+    if isinstance(value, bool):
+        return value
+    return AutoSetting(value)
```
Comment (hellkonig) [inline src/arxiv_recommender/embedding_config.py]: The `NormalizeEmbeddings` is just a boolean. Why do you need it?

File: src/arxiv_recommender/embedding_config.py
Comment (hellkonig) [inline src/arxiv_recommender/embedding_config.py:1]: Is it the best practise of top tier software engineering and ML engineering teams?

**Comment by hellkonig** [comment]
Addressing the latest comments on normalize_embeddings:

- The resolved runtime value is a boolean, but the requested user setting is not only a boolean: it can be true, false, or auto.
- I kept the requested-vs-resolved config split because this is the pattern I would expect in a mature ML system: user config can be ergonomic, then the application resolves it into explicit reproducible runtime settings for caching, logging, and evaluation.
- To make the distinction clearer, I renamed NormalizeEmbeddings to NormalizeEmbeddingsSetting in 70eda6f. ResolvedEmbeddingConfig.normalize_embeddings remains the concrete boolean used at runtime.

Local verification after the change:
- uv run ruff check .
- uv run ruff format --check .
- uv run python -m mypy src
- uv run python -m pytest
